### ISSUE TITLE
NotAny(FollowedBy()) failing

### ISSUE THREAD

**Issue by genry718** [pyparsing/pyparsing#482]
i have some input data like

`[gog1] [G1] [gog2] [gog3] [gog4] [G2] [gog5] [G3] [gog6]`

and want to find all gogs, if no G after it. so in this case i want to get `gog2` and `gog3` (and maybe `gog6`).

looks pretty simple, rigth? but i failed :(

```python
import pyparsing as pp
from pyparsing import *

def pyparsing_test():
    # this also dont helps
    # ParserElement.enable_left_recursion(force=True)

    data=""" [gog1] [G1] [gog2] [gog3] [gog4] [G2] [gog5] [G3] [gog6] """

    poi_type = Word(alphas).set_results_name('type')
    poi = Suppress('[') + poi_type + Char(nums) + Suppress(']')

    def cnd_is_type(val):
        return lambda toks: toks.type==val

    def cnd_is_not_type(val):
        return lambda toks: toks.type!=val

    poi_gog=poi('gog').add_condition(cnd_is_type('gog'))
    poi_g=poi('g').add_condition(cnd_is_type('G'))
    poi_not_g=poi('not_g').add_condition(cnd_is_not_type('G'))

    pattern = poi_gog + ~poi_g
    #WTF this finds only `gog6`, why??

    pattern = poi_gog + NotAny(FollowedBy(poi_g))
    #WTF same, only `gog6`

    pattern = poi_gog + poi_not_g.suppress()
    #WTF this works better but find only `gog2`, why not `gog3` also?

    r=pattern.search_string(data)
    print(data)
    print('='*10)
    print(r)
```

so my research shows that `NotAny(FollowedBy(poi_g))` not works at all. what im doing wrong?

**Comment by ptmcg** [comment]
I saw your post on SO, and have looked into it  a bit. Part of the problem is that results names will ordinarily be treated like dict keys - the last value assigned to a key will overwrite all previous values. Setting expression names and set_debug() will print out more of the step-by-step details of how the parser is working. As a short-cut, you can add these lines before calling search_string:

```python
    pp.enable_diag(pp.Diagnostics.enable_debug_on_named_expressions)
    pp.autoname_elements()
```

Changing your last pattern to:

```python
    pattern = poi_gog + FollowedBy(Group(poi_not_g) | StringEnd())
```

Gives the gog2, gog3, gog6 result.

I haven't had time to formulate a full response to your question, but maybe some of these hints might help you work through more of the issue on your own. If you _do_ find a bug in NotAny or FollowedBy, I'll be interested to hear about it!

**Comment by ptmcg** [comment]
Yes, this was a bug in NotAny, where parse actions were not being run. In your case a condition that returned False was not being run, so NotAny was erroneously thinking that the expression was a match, and therefore failing.

Fix is in progress, thanks for raising this!

**Comment by ptmcg** [comment]
You should be able to try it out in the current 3.1.0b2 release. Do `pip install -U pyparsing==3.1.0b2`

**Comment by ptmcg** [comment]
Closing - included in 3.1.0 release
